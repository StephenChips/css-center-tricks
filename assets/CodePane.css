.codepane {
    display: block;
}

.codepane__example {
    display: none;
    height: 100%;
}

.codepane__example.actived {
    display: block;
}

.codepane__example .codepane__code,
.codepane__example .codepane__result {
    display: none;
    height: 100%;
}

.codepane__example .codepane__result.actived {
    display: block;
}

.codepane__example .codepane__code.actived {
    display: flex;
}

.codepane__example .codepane__html,
.codepane__example .codepane__css {
    overflow: auto;
    height: 100%;
}

.codepane__example pre,
.codepane__example code {
    margin: 0;
    padding: 0;
    height: 100%;
    font-family: 'Fira Mono', 'Consolas', 'Courier New', Courier, monospace;
}

.codepane__example  .gutter.gutter-horizontal {
    cursor: ew-resize;
    border-left: dashed 1px lightgray;
    border-right: dashed 1px lightgray;
}
