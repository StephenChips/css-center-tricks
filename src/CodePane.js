import '../assets/CodePane.css';
import split from 'split.js';

// This is the key that we set on a style tag. It is used to identify
// the example which it relates to. 
const STYLE_EXAMPLE_KEYNAME = 'data-codepane-example-key';

// This is the key that we set on an example's element. It is the
// unique ID of the example in a CodePane.
const EXAMPLE_KEYNAME = 'data-key';

class CodePane {
    /**
     * 
     * @param {Element} el HTML Element of CodePane
     */
    constructor (el) {
        // There are two states for a CodePane, which are the 'showing-code' state and the 'showing-result' state.
        // When it's showing the current example's HTML & CSS snippets,  it's in the 'showing-code' state.
        // When it's showing the current example's result HTML, it's in the 'showing-result' state.
        this._state = 'showing-code';
        this._el = el;
        this._currentExampleKey = null;

        if (this._el.children.length > 0) {
            let firstChild = this._el.children[0];
            let keyOfFirstExample = firstChild.getAttribute(EXAMPLE_KEYNAME);

            // Initialize split.js for all examples.
            this._initSplitViewForAllExamples();

            // Calling the "showExample(...)" will ends the "creating" state.
            this.showExample(keyOfFirstExample);
        }
    }

    showExample (exampleKey) {
        if (exampleKey === this._currentExampleKey) {
            return;
        }

        let newActivedExample = this._findExampleByKey(exampleKey);
        if (newActivedExample === null) {
            throw new Error('The pane does not exists');
        }

        // If the codepane is under construction, or if there are 
        // no examples in an codepane, `this._currentExampleKey` will be null.
        if (this._currentExampleKey !== null) {
            let oldActivedExample = this._findExampleByKey(this._currentExampleKey);
            oldActivedExample.classList.remove('actived');
        }

        newActivedExample.classList.add('actived');
        this._currentExampleKey = newActivedExample.getAttribute(EXAMPLE_KEYNAME);

        const exampleState = this._getExampleState(newActivedExample);
        if (exampleState === 'showing-code' && this._state === 'showing-result') {
            this.showResult();
        } else if (exampleState === 'showing-result' && this._state === 'showing-code') {
            this.showCode();
        }
    }

    showCode () {
        if (this._el.children.length > 0) {
            // this._currentExampleKey points to the example that is currently being shown.
            // And because of that, it will not be null or undefined, except there are no
            // examples in the CodePane. In that situation, it must be null.
            let example = this._findExampleByKey(this._currentExampleKey);
            let code = example.querySelector('.codepane__code');
            let result = example.querySelector('.codepane__result');
    
            result.classList.remove('actived');
            code.classList.add('actived');
        }

        this._state = 'showing-code';
    }

    showResult () {
        if (this._el.children.length > 0) {
            // this._currentExampleKey points to the example that is currently being shown.
            // And because of that, it will not be null or undefined, except there are no
            // examples in the CodePane. In that situation, it must be null.
            let example = this._findExampleByKey(this._currentExampleKey);
            let code = example.querySelector('.codepane__code');
            let result = example.querySelector('.codepane__result');
    
            result.classList.add('actived');
            code.classList.remove('actived');    
        }

        this._state = 'showing-result';
    }

    /**
     * Add an example to a codepane.
     * @param {string} key example's key
     * @param {{ html: string, css: string, htmlSnippet: string, cssSnippet: string }}} param1 object specifies an example's content
     */
    addExample (key, { html, css, htmlSnippet, cssSnippet }) {
        if (this._findExampleByKey(key) !== null) {
            throw new Error('Your example\'s key is same with an existed example.');
        }

        // Firstm we create DOM structure through arguments.
        const $example = this._parseHTML(
`<div class="codepane__example" data-key="${key}">
    <div class="codepane__code">
        <div class="codepane__html">
            <pre><code>${htmlSnippet}</code></pre>
        </div>
        <div class="codepane__css">
            <pre><code>${cssSnippet}</code></pre>
        </div>
    </div>
    <div class="codepane__result">
        ${html}
    </div>
</div>`
        );

        // Then, we initialize the CodePane's states. The code snippets is show first by default.
        const $result = $example.querySelector('.codepane__result');
        const $code = $example.querySelector('.codepane__code');
        if (this._state === 'showing-code') {
            $code.classList.add('actived');
        } else if (this._state === 'showing-result') {
            $result.classList.add('actived');
        }

        // Then, we insert <style> tag with scoped style rules to the <head> element.
        const styleTag = document.createElement('style');
        styleTag.setAttribute(STYLE_EXAMPLE_KEYNAME, key);
        styleTag.textContent = this._addScopeToCssRules(css);
        
        this._el.appendChild($example);
        document.head.appendChild(styleTag);

        // Finally, initalize split.js for this new example.
        this._initSplitViewForExample($example);
    }

    _getExampleState (example) {
        var elCode = example.querySelector('.codepane__code');
        if (elCode.classList.contains('actived')) {
            return 'showing-code';
        } else {
            return 'showing-result';
        }
    }

    _addScopeToCssRules (css) {
        return css;
    }

    _parseHTML (str) {
        let wrapper = document.createElement('div');
        wrapper.innerHTML = str;
        return wrapper.firstElementChild;
    }

    _createScopedCSSObject (key, styleObject) {
        // We assume all rules in the style object are not scoped.
        let scopedCSSRules = Object.create(null);
        
        for (let selector in styleObject) {
            let scopedSelector = `.codepane__example[data-key="${key}"] ${selector}`;
            scopedCSSRules[scopedSelector] = Object.assign({}, styleObject[selector])
        }
        
        return scopedCSSRules;
    }

    _convertCSSObjectToString (styleObject) {
        let result = '';

        for (let selector in styleObject) {
            let styleRules = styleObject[selector];
            let strRules  = '';

            for (let property in styleRules) {
                let valueOfProperty = styleRules[property];
                strRules += `${property}: ${valueOfProperty};`;
            }

            result += `${selector} { ${strRules} }`;
        }

        return result;
    }

    _findExampleByKey (key) {
        return this._el.querySelector(`.codepane__example[data-key="${key}"]`);
    }

    _initSplitViewForExample (example) {
        let htmlView = example.querySelector('.codepane__code .codepane__html');
        let cssView = example.querySelector('.codepane__code .codepane__css');

        if (htmlView !== null && cssView !== null) {
            split([ htmlView, cssView ], {
                gutterSize: 5,
                elementStyle (dimension, size, gutterSize) {
                    return {
                        'flex-basis': 'calc(' + size + '% - ' + gutterSize + 'px)',
                    }
                },
                gutterStyle (dimension, gutterSize) {
                    return {
                        'flex-basis': gutterSize + 'px',
                    }
                },
            });
        }
    }

    _initSplitViewForAllExamples () {
        let allExamples = this._el.querySelectorAll('.codepane__example');
        for (let example of allExamples) {
            this._initSplitViewForExample(example);
        }
    }

    get el () {
        return this._el;
    }
}

export default CodePane;
