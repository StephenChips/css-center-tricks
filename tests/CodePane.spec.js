import CodePane from '../src/CodePane';
import { tick } from './test-utils';

/**
 * There are three aspects for testing a component:
 * 1. Behaviour
 * 2. Element structure
 * 3. Style
 * 
 * For testing component's behaviour, you can alter initialize your component, control the input and check for output.
 * Input is things will make the component changed, and an output. is a things that is effected by the component's change. 
 * 
 * Common inputs include:
 *    1. calling component's method.
 *    2. setting component's property/data.
 *    3. triggering a DOM event
 * 
 * Common outputs include:
 *    1. DOM structure's change
 *    2. component's property/data's change
 *    3. recieving a event and calling the callback.
 * 
 * Inputs and output may be various, many of them are actually the same, but they should be treated differently. 
 * You should analyze and find them out before writing a test.
 * 
 * 2. For testing component's element structure, you can use Jest's snapshot testing. you can set to component to a certain state first, then take a snapshot.
 * 
 * 3. For testing component's style,  you can do the similiar thing as testing behaviour.
 *    You can observe the style changes by calling getComputedStyle(...) method.
 */

const CODEPANE_WITH_TWO_EXAMPLES = `
<div id="codepane" class="codepane">
    <div class="codepane__example" data-key="A">
        <div class="codepane__result"></div>
        <div class="codepane__code"></div>
    </div>
    <div class="codepane__example" data-key="B">
        <div class="codepane__result"></div>
        <div class="codepane__code"></div>
    </div>
</div>
`;

const CODEPANE_WITH_NO_EXAMPLES = `<div id="codepane" class="codepane"></div>`;

function createCodePane (template) {
    document.body.innerHTML = template;
    return new CodePane(document.getElementById('codepane'));
}

describe('Test initialization, showExample, showCode and showResult', () => {
    // We used the CodePane template with two examples for testing
    // (see above). For convenient, we declear following variables,
    // which will be assigned to correspondent HTML element
    // before each test starts.

    let firstExample; // it points to the first example's element.
    let secondExample; // it points to the second example's element.

    // it points to the first example's code snippet pane's element
    // (the one shows after calling "showCode(...)").
    let firstCode; 
    
    // it points to the first example's result pane's element
    // (the one shows after calling "showResult(...)").
    let firstResult;

    // it points to the first example's code snippet pane's element
    // (the one shows after calling "showCode(...)").
    let secondCode;

    // it points to the second example's result pane's element
    // (the one shows after calling "showResult(...)").
    let secondResult;

    let codepane; // the CodePane instance that we are going to test.
    let el; // HTML elements controled by testing CodePane.

    function setupElements () {
        // We assumes that the codepane has been initialized. and its element is accessible.
        el = codepane.el;
        firstExample = el.children[0];
        secondExample = el.children[1];
        firstCode = firstExample.querySelector('.codepane__code');
        firstResult = firstExample.querySelector('.codepane__result');
        secondCode = secondExample.querySelector('.codepane__code');
        secondResult = secondExample.querySelector('.codepane__result');
    }

    it('will shows the first example by default', () => {
        codepane = createCodePane(CODEPANE_WITH_TWO_EXAMPLES);

        setupElements();

        expect(firstExample.classList.contains('actived')).toBe(true);
        expect(secondExample.classList.contains('actived')).toBe(false);
    });

    it('re-show the same example won\'t make any changes', () => {
        codepane = createCodePane(CODEPANE_WITH_TWO_EXAMPLES);

        setupElements();

        codepane.showExample('A');
        expect(firstExample.classList.contains('actived')).toBe(true);
        expect(secondExample.classList.contains('actived')).toBe(false)
    });

    it('showExample(\'B\')', () => {
        codepane = createCodePane(CODEPANE_WITH_TWO_EXAMPLES);

        codepane.showExample('B');

        setupElements();

        expect(firstExample.classList.contains('actived')).toBe(false);
        expect(secondExample.classList.contains('actived')).toBe(true);
        expect(secondCode.classList.contains('actived')).toBe(true);
        expect(secondResult.classList.contains('actived')).toBe(false);
    });

    it('throws errors if the example to show does not exists.', () => {
        expect(() => {
            codepane = createCodePane(CODEPANE_WITH_TWO_EXAMPLES);
            codepane.showExample('C');
        }).toThrow('The pane does not exists');
    });

    it('new CodePane(el) -> showResult()', () => {
        codepane = createCodePane(CODEPANE_WITH_TWO_EXAMPLES);
        codepane.showResult();

        setupElements();

        expect(secondCode.classList.contains('actived')).toBe(false);
        expect(secondResult.classList.contains('actived')).toBe(false)
    });

    // If the code's behaviour doesn't meet the requirement,
    // even all paths are covered, there are still bugs.
    it('If a CodePane is displaying the result HTML, when we switch to another example, it will display that example\'s result HTML.', async () => {
        codepane = createCodePane(CODEPANE_WITH_TWO_EXAMPLES);
        codepane.showExample('B');
        codepane.showResult();
        await tick();
        codepane.showExample('A');

        setupElements();

        expect(firstResult.classList.contains('actived')).toBe(true);
        expect(firstCode.classList.contains('actived')).toBe(false);
    });
});

describe('Test addExample(...)', () => {
    const NEW_EXAMPLE_KEY = 'dummy';

    const html = '<div></div>'
    const css = `div { display: flex; }`;
    const scopedCss = `.codepane__example[data-example-key="${NEW_EXAMPLE_KEY}"] div { display: flex }`;
    const htmlSnippet = '&lt;div&gt;&lt;div/&gt;';
    const cssSnippet = [
        '<span class="selector">div</span>&nbsp;{',
        '    <span class="property-name">display</span>:&nbsp;<span class="property-value">flex</span>;',
        '}',
    ].join('<br/>');

    let codepane, el;

    async function createCodePaneAndAddOneExample () {
        codepane = createCodePane(CODEPANE_WITH_NO_EXAMPLES);
        codepane.addExample(NEW_EXAMPLE_KEY, {
            html,
            css,
            htmlSnippet,
            cssSnippet
        });

        await tick();

        el = codepane.el;
        return codepane;
    }
    
    it('render result correctly when we call "showResult()"', async () => {
        await createCodePaneAndAddOneExample();

        codepane.showExample(NEW_EXAMPLE_KEY);
        codepane.showResult();

        expect(codepane.el).toMatchSnapshot();
    });

    it('render HTML & CSS snippets correctly when we call "showCode()"', async () => {
        await createCodePaneAndAddOneExample();

        codepane.showExample(NEW_EXAMPLE_KEY);
        codepane.showCode();

        expect(codepane.el).toMatchSnapshot();
    });

    it('cannot add two example with same key', async () => {
        const addSameExampleTwice = async () => {
            codepane.addExample('SAME_KEY', {
                html,
                css,
                htmlSnippet,
                cssSnippet
            });

            await tick();
            codepane.addExample('SAME_KEY', {
                html,
                css,
                htmlSnippet,
                cssSnippet
            });

            await tick();
        };

        await expect(addSameExampleTwice()).rejects.toBeInstanceOf(Error);
    });
    
    it('will create example\'s element with the correct key and append it to CodePane.', async () => {
        await createCodePaneAndAddOneExample();
        const elExample = el.querySelector(`.codepane__example[data-key="${NEW_EXAMPLE_KEY}"]`);
        expect(elExample).not.toBeNull();
    });

    it('The new example will not automatically become actived until we call "addExample(...)".', async () => {
        await createCodePaneAndAddOneExample();

        // First, we select the example's element here. It should not be actived by default.
        // expect(...).toBeNull() means we cannot find the actived example.
        expect(el.querySelector(`.codepane__example[data-key="${NEW_EXAMPLE_KEY}"].actived`)).toBeNull(); 

        // Until we call "showExample(...)".
        codepane.showExample(NEW_EXAMPLE_KEY);
        // expect(...).not.toBeNull() means we found the actived example.
        expect(el.querySelector(`.codepane__example[data-key="${NEW_EXAMPLE_KEY}"].actived`)).not.toBeNull(); 
    });

    it('the new example shows code snippets if CodePane\'s current state is \'showing-code\'', async () => {
        let codepane = createCodePane(CODEPANE_WITH_NO_EXAMPLES);

        codepane.showCode();

        await tick();
    
        codepane.addExample('EXAMPLE_ONE', {
            html,
            css,
            htmlSnippet,
            cssSnippet
        });

        await tick();

        const elCode = codepane.el.querySelector(`.codepane__example[data-key="EXAMPLE_ONE"] .codepane__code`);
        const elResult = codepane.el.querySelector(`.codepane__example[data-key="EXAMPLE_ONE"] .codepane__result`);
        expect(elCode.classList.contains('actived')).toBe(true);
        expect(elResult.classList.contains('actived')).toBe(false);
    });

    it('the new example shows the result HTML if CodePane\'s current state is \'showing-result\'', async () => {
        let codepane = createCodePane(CODEPANE_WITH_NO_EXAMPLES);

        codepane.showResult();

        await tick();
    
        codepane.addExample('EXAMPLE_ONE', {
            html,
            css,
            htmlSnippet,
            cssSnippet
        });

        await tick();

        const elCode = codepane.el.querySelector(`.codepane__example[data-key="EXAMPLE_ONE"] .codepane__code`);
        const elResult = codepane.el.querySelector(`.codepane__example[data-key="EXAMPLE_ONE"] .codepane__result`);
        expect(elCode.classList.contains('actived')).toBe(false);
        expect(elResult.classList.contains('actived')).toBe(true);
    });

    // [FAIL] Bacause we doesn't add name to the style tag
    it('will add a <style> tag in the <head> element', async () => {
        await createCodePaneAndAddOneExample();

        let styleTag = document.querySelector(`head > style[data-codepane-example-key="${NEW_EXAMPLE_KEY}"]`);
        expect(styleTag).toBeInstanceOf(HTMLStyleElement);

        // the style rules' selector will be scoped so that it can only effect elements inside a specific example.
        expect(styleTag.textContent).toBe(scopedCss);
    });
});
