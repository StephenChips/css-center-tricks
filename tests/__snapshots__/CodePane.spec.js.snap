// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test addExample(...) render HTML & CSS snippets correctly when we call "showCode()" 1`] = `
<div
  class="codepane"
  id="codepane"
>
  <div
    class="codepane__example actived"
    data-key="dummy"
  >
    
    
    <div
      class="codepane__code actived"
    >
      
        
      <div
        class="codepane__html"
        style="flex-basis: calc(50% - 2.5px);"
      >
        
            
        <pre>
          <code>
            &lt;div&gt;&lt;div/&gt;
          </code>
        </pre>
        
        
      </div>
      
        
      <div
        class="gutter gutter-horizontal"
        style="flex-basis: 5px;"
      />
      <div
        class="codepane__css"
        style="flex-basis: calc(50% - 2.5px);"
      >
        
            
        <pre>
          <code>
            <span
              class="selector"
            >
              div
            </span>
             {
            <br />
                
            <span
              class="property-name"
            >
              display
            </span>
            : 
            <span
              class="property-value"
            >
              flex
            </span>
            ;
            <br />
            }
          </code>
        </pre>
        
        
      </div>
      
    
    </div>
    
    
    <div
      class="codepane__result"
    >
      
        
      <div />
      
    
    </div>
    

  </div>
</div>
`;

exports[`Test addExample(...) render result correctly when we call "showResult()" 1`] = `
<div
  class="codepane"
  id="codepane"
>
  <div
    class="codepane__example actived"
    data-key="dummy"
  >
    
    
    <div
      class="codepane__code"
    >
      
        
      <div
        class="codepane__html"
        style="flex-basis: calc(50% - 2.5px);"
      >
        
            
        <pre>
          <code>
            &lt;div&gt;&lt;div/&gt;
          </code>
        </pre>
        
        
      </div>
      
        
      <div
        class="gutter gutter-horizontal"
        style="flex-basis: 5px;"
      />
      <div
        class="codepane__css"
        style="flex-basis: calc(50% - 2.5px);"
      >
        
            
        <pre>
          <code>
            <span
              class="selector"
            >
              div
            </span>
             {
            <br />
                
            <span
              class="property-name"
            >
              display
            </span>
            : 
            <span
              class="property-value"
            >
              flex
            </span>
            ;
            <br />
            }
          </code>
        </pre>
        
        
      </div>
      
    
    </div>
    
    
    <div
      class="codepane__result actived"
    >
      
        
      <div />
      
    
    </div>
    

  </div>
</div>
`;
